### PR TITLE
[MRG] Add ERDS maps to time-frequency examples

### LINKED ISSUE #4558: Custom colormap normalization in power plot
I'd like to create a classic ERD/ERS time/frequency plot by computing a TFR (e.g. via `tfr_multitaper`). Power should be normalized to a baseline with `mode='percent'`. I want to set the limits to `vmin=-1` and `vmax=1.5` using `cmap='RdBu'` (because these are the values typically used in ERD/ERS maps). This works fine, but now I'd like to map the value 0 to white; negative values should linearly scale from -1 (red) to 0 (white), and positive values should linearly scale from 0 (white) to 1.5 (blue).

I found [this example](https://matplotlib.org/users/colormapnorms.html#custom-normalization-two-linear-ranges) on how to normalize colormaps using the `norm` argument of `pcolormesh` (the relevant line in MNE that produces the plot seems to be [here](https://github.com/mne-tools/mne-python/blob/master/mne/viz/topo.py#L309)). Is it possible to do this after the plot has been created (if so, how)? Or is this only possible by specifying the `norm` argument for `pcolormesh`?

### PR DESCRIPTION
**Pull request by cbrnr**
Fix #4558. This example produces the following two plots:

figure_1
figure_2

This is already pretty OK, but there are two minor things that I'd like to change if it's not too much effort:

1. The x-axis should be in seconds.
2. The y-axis is weirdly spaced, ideally I'd like the frequency tick labels to be uniformly spaced.

Any easy ways to fix these things?

### CODE REVIEW

**Comment by larsoner** [comment]
> The x-axis should be in seconds.

Agreed, we've converged on using sec almost everywhere

> The y-axis is weirdly spaced, ideally I'd like the frequency tick labels to be uniformly spaced.

This is debatable. IIRC the current behavior puts frequency ticks at centers that are actually used. What you propose would not.

**Comment by larsoner** [comment]
... so for the latter, you can do it in `mpl` if you prefer the even spacing (`for ax in fig.axes: ax.set_yticks(np.linspace(*ax.get_ylim(), n=10, endpoint=True))`) but I'd rather not change it

File: examples/time_frequency/plot_time_frequency_erds.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    index_new = np.hstack([np.linspace(0, vzero, cmap.N // 2, endpoint=False),
+                           np.linspace(vzero, 1, cmap.N // 2)])
+    cdict = {"red": [], "green": [], "blue": [], "alpha": []}
+    for old, new in zip(index_old, index_new):
+        r, g, b, a = cmap(old)
+        cdict["red"].append((new, r, r))
+        cdict["green"].append((new, g, g))
+        cdict["blue"].append((new, b, b))
+        cdict["alpha"].append((new, a, a))
+    return LinearSegmentedColormap("erds", cdict)
```
Comment (larsoner) [inline examples/time_frequency/plot_time_frequency_erds.py]: This solution seems less preferable than going the `Norm` class route. Why not use the norm class that we discussed and use `pc.set_norm(...)`?

Comment (cbrnr) [reply examples/time_frequency/plot_time_frequency_erds.py]: Yes, I just had this example almost implemented with this quick fix. I'll try the other approach and see if it works.

**Comment by cbrnr** [comment]
Regarding having the x-axis in seconds, should I change the underlying function or just edit the labels after the fact for this example?

I don't understand what you mean by centers that are actually used? I'm using frequencies from 2 to 31.5Hz in steps of 0.5 (I actually wanted to use steps of 1Hz but forgot to revert this, but this doesn't change the behavior). All frequencies are used, but why are the labels unevenly spaced? They could just as well be evenly spaced, right? In this example with 0.5Hz steps, the only exception is the jump from 7 to 10Hz, all other intervals are 2.5Hz.

**Comment by larsoner** [comment]
Because it only uses some maximum number from the original, maybe ten. Feel
free to propose an alternative but keep in mind some use cases could use
hundreds of frequencies so we can't just use all.

For example maybe we should choose whatever the smallest decimation (slice)
is that results in ten or fewer frequency labels. This would be evenly
spaced (advantage) but wouldn't necessarily contain both bounds
(disadvantage).

**Comment by agramfort** [comment]
see how it looks like:

https://5700-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.1dFPtus/html/auto_examples/time_frequency/plot_time_frequency_erds.html

you have edge artifacts (see red bars on edges)

**Comment by cbrnr** [comment]
@agramfort `psd_tfr_multitaper` doesn't have `tmin` and `tmax` arguments - how do I remove these edge artifacts?

**Comment by cbrnr** [comment]
@larsoner ah, you mean the start and stop frequencies are always included. This makes sense, but I still find non-equally spaced labels a bit strange. I like your suggestion, we could always include the lower bound and then find the smallest decimation that results in ten labels (but in another PR).

What about the x axis in s? I guess this should also be changed in a separate PR?

**Comment by larsoner** [comment]
Probably easier to change the aesthetics of existing functions separately,
especially as it increases the probability of disagreement and bike shedding

**Comment by cbrnr** [comment]
I tried using the class we discussed earlier, but changing the normalization of an existing image is not sufficient; the colorbar needs be re-normalized as well. The former can be achieved by

    ax[i].collections[0].set_norm(OffsetNorm(vmin=vmin, vmax=vmax, midpoint=0))

which is not really intuitive. Something along this line must be done for the colorbar too, but I haven't figured out how to do this.

Furthermore, I'm not sure how to define a class derived by `matplotlib.colors.Normalize` if MNE doesn't do top-level imports of matplotlib.

Bottom line is that I like the current solution with a custom colormap much more than using this renormalizing hack. So if you agree I'd like to keep it as is. Then the final thing to do in this PR is to get rid of the edge artifacts - any hints?

**Comment by larsoner** [comment]
Okay we can use the colormap trick. Might be worth linking to the norm stuff, too.

Compute over a larger window than you want to actually use, then do `tfr.crop(...)` to remove edge artifacts.

**Comment by cbrnr** [comment]
All done. Thanks @larsoner!

File: examples/time_frequency/plot_time_frequency_erds.py [hunk 2]
```diff
@@ -47,8 +51,12 @@
...
 
 def center_cmap(cmap, vmin, vmax):
-    """Center given colormap at value 0.
-    """
+    """Center given colormap (ranging from vmin to vmax) at value 0.
+
+    Note that eventually this could also be achieved by re-normalizing a given
+    colormap by subclassing matplotlib.colors.Normalize as described here:
+    https://matplotlib.org/users/colormapnorms.html#custom-normalization-two-linear-ranges
+    """  # noqa 501
```
Comment (larsoner) [inline examples/time_frequency/plot_time_frequency_erds.py]: This will probably `noqa` everything, I think you mean `# noqa: E501`

**Comment by cbrnr** [comment]
Ups, fixed.

**Comment by larsoner** [comment]
Can you paste the CircleCI link to your new example once it's up? Then @agramfort can review and merge if he's happy

**Comment by cbrnr** [comment]
Sure, here goes:
https://5749-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.Ve5BZpx/html/auto_examples/time_frequency/plot_time_frequency_erds.html

**Comment by agramfort** [comment]
is it possible to have colorbar on the plots?

**Comment by cbrnr** [comment]
Yeah, good question. Doesn't MNE have something to make room for a colorbar on a given figure? Otherwise, I guess I'll have to dig into MPL again.

**Comment by agramfort** [comment]
see for example how ERP image layout is done

**Comment by cbrnr** [comment]
@agramfort https://5760-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.Xb783V7/html/auto_examples/time_frequency/plot_time_frequency_erds.html

**Comment by larsoner** [comment]
LGTM +1 for merge

We can fix the `whats_new.rst` conflict locally nowadays so no need to rebase just for that. The only CI failure was flake which you fixed so again let's assume CIs are happy

**Comment by cbrnr** [comment]
OK. Is it just me or is Travis getting slower and slower? Average time for a commit seems to be over 2 hours now.

**Comment by larsoner** [comment]
It depends not just on the Travis service, but also how many PRs are open and being merged. (Each merged PR gets run in full, again, after merge.)

**Comment by cbrnr** [comment]
Yes, but the resources of the Travis service seem to operate at their limits. It's the wait times before a job actually starts that really got very long, and I remember that wasn't always the case.

**Comment by larsoner** [comment]
> It's the wait times before a job actually starts that really got very long

That's what I'm talking about actually. We only get 5 (or 4?) concurrent jobs for the `mne-tools` organization. Each commit uses 9 jobs (each lasting 10-20 minutes). Every open PR (to MNE-Python or MNE-CPP) adds these jobs to the queue, as do any merges.

Go to the [Travis status page for mne-python](https://travis-ci.org/mne-tools/mne-python/pull_requests) and you'll see a lot of yellow, with the bottom-most yellow one(s) actually running.

**Comment by larsoner** [comment]
... but don't forget that anything you really wait for the CIs for you shoudl be able to do locally. So for example a habit of ding `make pep` before pushing a commit can often save a 2-hour disappointment :)

**Comment by larsoner** [comment]
(Back in the day when mne-python started we didn't even have CIs so we had to run tests ourselves -- like a horrible sort of distributed mechanical turk CI system)

**Comment by cbrnr** [comment]
Yeah that's true, I should really do more tests locally...

**Comment by agramfort** [comment]
+1 for MRG when CIs are green and conflicts are fixed

thx @cbrnr

**Comment by cbrnr** [comment]
@larsoner should I rebase or do you take care of that?

**Comment by larsoner** [comment]
I will use LarsonCI™ to test that you fixed the one error you had, fix the conflict, then merge

**Comment by larsoner** [comment]
thanks @cbrnr
